File: VB/HostApp/ReportService.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Reflection
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Parameters

Namespace HostApp
	Public Class ReportService
		Implements IReportService
		Public Function GetReportNames() As String() Implements IReportService.GetReportNames
			Dim types() As Type = System.Reflection.Assembly.GetExecutingAssembly().GetTypes()
			Dim reportTypes As New List(Of String)()

			For i As Integer = 0 To types.Length - 1
				If GetType(XtraReport).IsAssignableFrom(types(i)) Then
					reportTypes.Add(types(i).FullName)
				End If
			Next i

			Return reportTypes.ToArray()
		End Function

		Public Function GetReportLayout(ByVal reportName As String) As Byte() Implements IReportService.GetReportLayout
			Dim report As XtraReport = CreateReportByName(reportName)
			Dim memoryStream As New MemoryStream()

			report.SaveLayout(memoryStream)

			Return memoryStream.ToArray()
		End Function

		Public Function GetReportDocument(ByVal reportName As String) As Byte() Implements IReportService.GetReportDocument
			Dim report As XtraReport = CreateReportByName(reportName)
			Dim memoryStream As New MemoryStream()

			' Disable report parameters' visibility to prevent the 'Parameters' dialog from popping up in the WCF host process
			For i As Integer = 0 To report.Parameters.Count - 1
				report.Parameters(i).Visible = False
			Next i

			report.CreateDocument()
			report.PrintingSystem.SaveDocument(memoryStream)

			Return memoryStream.ToArray()
		End Function

		Public Function GetParametrizedReportDocument(ByVal reportName As String, ByVal parameters() As ReportParameterInfo) As Byte() Implements IReportService.GetParametrizedReportDocument
			Dim report As XtraReport = CreateReportByName(reportName)
			Dim memoryStream As New MemoryStream()

			' Disable report parameters' visibility to prevent the 'Parameters' dialog from popping up in the WCF host process
			For i As Integer = 0 To report.Parameters.Count - 1
				report.Parameters(i).Visible = False
			Next i

			For i As Integer = 0 To parameters.Length - 1
				Dim repPar As Parameter = report.Parameters(parameters(i).Name)

				If repPar IsNot Nothing Then
					repPar.Value = parameters(i).Value
				End If
			Next i

			report.CreateDocument()
			report.PrintingSystem.SaveDocument(memoryStream)

			Return memoryStream.ToArray()
		End Function

		Private Function CreateReportByName(ByVal reportName As String) As XtraReport
			Return CType(Activator.CreateInstance(Nothing, reportName).Unwrap(), XtraReport)
		End Function
	End Class
End Namespace

File: CS/HostApp/ReportService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Parameters;

namespace HostApp {
    public class ReportService : IReportService {
        public string[] GetReportNames() {
            Type[] types = Assembly.GetExecutingAssembly().GetTypes();
            List<string> reportTypes = new List<string>();

            for (int i = 0; i < types.Length; i++) {
                if (typeof(XtraReport).IsAssignableFrom(types[i]))
                    reportTypes.Add(types[i].FullName);
            }

            return reportTypes.ToArray();
        }

        public byte[] GetReportLayout(string reportName) {
            XtraReport report = CreateReportByName(reportName);
            MemoryStream memoryStream = new MemoryStream();

            report.SaveLayout(memoryStream);

            return memoryStream.ToArray();
        }

        public byte[] GetReportDocument(string reportName) {
            XtraReport report = CreateReportByName(reportName);
            MemoryStream memoryStream = new MemoryStream();

            // Disable report parameters' visibility to prevent the 'Parameters' dialog from popping up in the WCF host process
            for (int i = 0; i < report.Parameters.Count; i++) {
                report.Parameters[i].Visible = false;
            }

            report.CreateDocument();
            report.PrintingSystem.SaveDocument(memoryStream);

            return memoryStream.ToArray();
        }

        public byte[] GetParametrizedReportDocument(string reportName, ReportParameterInfo[] parameters) {
            XtraReport report = CreateReportByName(reportName);
            MemoryStream memoryStream = new MemoryStream();

            // Disable report parameters' visibility to prevent the 'Parameters' dialog from popping up in the WCF host process
            for (int i = 0; i < report.Parameters.Count; i++) {
                report.Parameters[i].Visible = false;
            }

            for (int i = 0; i < parameters.Length; i++) {
                Parameter repPar = report.Parameters[parameters[i].Name];
                
                if (repPar != null)
                    repPar.Value = parameters[i].Value;
            }

            report.CreateDocument();
            report.PrintingSystem.SaveDocument(memoryStream);

            return memoryStream.ToArray();
        }

        private XtraReport CreateReportByName(string reportName) {
            return (XtraReport)Activator.CreateInstance(null, reportName).Unwrap();
        }
    }
}